File: 2023/HKCERT/ST Code I and II/solve.py
import aiohttp
import asyncio
import aiofiles
from lxml import etree
import qrcode
import requests
import qrcode.image.svg
from bs4 import BeautifulSoup

URL = "http://stcode-3983gi.hkcert23.pwnable.hk:28211"
session: aiohttp.ClientSession


async def generateQR(data: str) -> str:
    """
    Generates a QR code as SVG.
    """
    factory = qrcode.image.svg.SvgImage
    # qr = qrcode.QRCode(
    #     version=1,
    #     error_correction=qrcode.constants.ERROR_CORRECT_L,
    #     box_size=10,
    #     border=4,
    #     image_factory=factory,
    # )
    # qr.add_data(data)
    # qr.make(fit=True)
    # img = qr.make_image(fill_color="black", back_color="white")
    #
    # image = img.get_image()

    img = qrcode.make(data, image_factory=factory)
    image = img.get_image()

    return etree.tostring(image).decode()


async def encodeST(st: str, qr_code: BeautifulSoup):
    """
    Encodes the given ST string into the QR code in binary.
    """
    binary = bin(int.from_bytes(st.encode(), "big"))[2:]
    print(binary)


async def decode_ST(svg: BeautifulSoup):
    # For each rect, check if rx is set, if it is read it into a list
    rx = ''
    for rect in svg.find_all("rect"):
        if "rx" in rect.attrs:
            rx += rect["rx"]
    # We get a binary string
    # print(rx)

    # Decode binary string to ascii
    # print(int(rx, 2))
    st = bytes.fromhex(hex(int(rx, 2))[2:]).decode("utf-8")
    return st


async def flag1():
    async with session.get(URL + "/flag1") as resp:
        svg = await resp.text()
        soup = BeautifulSoup(svg, 'xml')
        # print(soup.prettify())
        flag_1 = await decode_ST(soup)
    #     async with aiofiles.open("files/flag1.svg", "w") as f:
    #         await f.write(svg)
    # output = await readQR("files/flag1.svg")
    return svg, flag_1


async def flag2(flag_1):
    async with session.get(URL + "/flag2") as resp:
        res_text = await resp.text()
        soup = BeautifulSoup(res_text, 'html.parser')
        # print(soup.prettify())
        # Send an svg that contains the QRCode of flag1
        # curl http://stcode-3983gi.hkcert23.pwnable.hk:28211/flag2 -F svg=@YOUR_PAYLOAD_FILE --cookie "connect.sid="


    # Send flag_1 as file
    flag_1_qr = await generateQR(flag_1)
    print(flag_1_qr)


    data = aiohttp.FormData()
    payload_file = "files/flag1.svg"
    async with aiofiles.open(payload_file, "w") as f:
        await f.write(flag_1_qr)
    data.add_field('svg', open(payload_file, 'rb'))

    async with session.post(URL + "/flag2", data=data) as resp:
        res_text = await resp.text()
        soup = BeautifulSoup(res_text, 'html.parser')
        print(soup.prettify())


    # Try using requests
    files = {"svg": (payload_file, open(payload_file, 'rb'))}
    cookies = {"connect.sid": session.cookie_jar.filter_cookies(URL).get("connect.sid").value}
    resp = requests.post(URL + "/flag2", files=files, cookies=cookies)
    print(resp.text)


async def main():
    # flag = "hkcert23{ST_ST&s4_STegan0graphy--STeg0}"
    # qr_code_xml = await generateQR(flag)
    # soup = BeautifulSoup(qr_code_xml, 'xml')
    # async with aiofiles.open("files/flag1.svg", "w") as f:
    #     await f.write(qr_code_xml)
    # print(soup.prettify())
    #
    # return

    global session
    session = aiohttp.ClientSession()

    flag_1_svg, flag_1 = await flag1()
    print(f"Flag 1: {flag_1}")
    flag_2 = await flag2(flag_1)

    await session.close()
    pass

if __name__ == "__main__":
    asyncio.run(main())

File: 2023/HKCERT/ST Code I and II/solve.js
// import fs from 'fs';
import fetch from "node-fetch";
import {FormData, Blob} from "node-fetch";
import QRCode from 'qrcode-svg';
import jsQR from 'jsqr';
import sharp from 'sharp';
import fs from 'fs';

const SERVER = 'http://stcode-3983gi.hkcert23.pwnable.hk:28211';
const flag1 = 'hkcert23{ST_ST&s4_STegan0graphy--STeg0}';
// const flag1ST = encodeST(flag1, encodeQR(flag0));

// encodeST and decodeST are based on RegExp
function encodeST(data, svg) {
    // Convert data to binary array
    const data_bin = data.split('').map(c => c.charCodeAt(0).toString(2).padStart(8, '0')).join('');
    console.log(data_bin);

    // Count number of rect elements in svg
    const num_rects = (svg.match(/<rect/g) || []).length;
    // If number of rects is not enough, add more
    if (num_rects < data_bin.length) {
        // First remove end tag
        svg = svg.replace('</svg>', '');
        // For each bit of data, add a rect element
        svg += '<rect x="0" y="0" width="0" height="0" fill="none" />';
        for (let i = 0; i < data_bin.length - num_rects; i++) {
            svg += '<rect x="0" y="0" width="0" height="0" fill="none" />';
        }
        // Add back end tag
        svg += '</svg>';
    }

    // Create the encoded svg by adding a rx attribute to all rect elements (except the first one)
    // each rx attribute is a bit of the data
    let i = -1;
    const encoded_svg = svg.replace(/<rect/g, (match, offset) => {
        i++;
        if (i === 0) {
            return match;
        }
        return `<rect rx="${data_bin[i - 1]}"`;
    });

    console.log(encoded_svg);

    return encoded_svg;
}

function decodeST(svg) {

}

function encodeQR(data) {
    return new QRCode(data).svg();
}

async function decodeQR(svg) {
    try {
        const {data, info} = await sharp(new Buffer(svg)).ensureAlpha().raw().toBuffer({resolveWithObject: true});
        const output = await jsQR(new Uint8ClampedArray(data.buffer), info.width, info.height);
        return output.data;
    } catch (e) {
        console.log(e);
        return null;
    }
}

async function uploadFile(file, cookies) {
    // Send POST request to flag2 with cookie and svg payload of flag1
    const data = new FormData();
    data.append("svg", new Blob([file]));
    const res2 = await fetch(SERVER + '/flag2', {
        method: 'POST',
        headers: {
            'Cookie': cookies,
        },
        body: data,
    });

    return await res2.text();
}

async function solve() {
    console.log("Getting session cookies...")
    // Send fetch request to flag2 to get cookie, do not allow redirects
    const res = await fetch(SERVER + "/flag2", {
        method: 'GET',
        redirect: 'manual'
    });
    // Get connect.sid cookie
    const cookies = res.headers.get('set-cookie');
    console.log("Cookies: " + cookies);

    console.log("Sending POST request to flag2...")

    const flag1_qr = encodeQR(flag1);
    // console.log(flag1_qr);

    // Save flag1 to file
    // fs.writeFileSync('files/flag1_2.svg', flag1_qr);

    // Try to decode
    const flag1_decoded = await decodeQR(flag1_qr);
    console.log("Decoded: " + flag1_decoded);

    // Send POST request to flag2 with cookie and svg payload of flag1
    let res_text = await uploadFile(flag1_qr, cookies);

    // Print response
    console.log(res_text);

    // Loop 15 times
    for (let i = 0; i < 15; i++) {
        // Get QRCode text ans STCode from res_text
        const qr_text = res_text.split("QRCode:")[1].split("STCode:")[0].trim();
        const st_text = res_text.split("STCode:")[1].split("Your have")[0].trim();

        // Generate QRCode from qr_text
        const qr = await encodeQR(qr_text);
        const st_encoded_qr = await encodeST(st_text, qr);

        // Upload ST-encoded QR to /flag2
        res_text = await uploadFile(st_encoded_qr, cookies);
        console.log(res_text);
    }
}

// Run solve (it's async)
solve();
